pkg/ipcache/ipcache.go

```diff
@@ -69,12 +70,37 @@ type IPKeyPair struct {
 	Key uint8
 }
 
+const (
+	rLockWeight = 1
+	lockWeight  = 2
+)
+
+type semaphoreRWMutex struct {
```

Comment:
I would love to see this in `pkg/lock` instead including proper unit tests

Reply:
@tgraf Done please take a look